### ISSUE TITLE
Mouse selection always chooses 1st in list

### ISSUE THREAD

**Issue by 0s1s** [bassjobsen/Bootstrap-3-Typeahead#331]
I have just upgraded to Firefox 54 and typeahead has stopped functioning where it pertains to selecting a value with the mouse. What is happening is any selection you make with the mouse will always result in the first value being selected. So, if you type "hal" and want haley, you'd click on haley but "hal Miller" is what's shown in the console and populated to the input.

Here is the code that will reproduce the issue. Also, the same behavior is occurring in Chrome and Opera.

```
  $("#search").typeahead({
    source: function(query, process) {
      var data = [
        {"name":"Hal Miller"},
        {"name":"Haley Jones"}
      ];

      return process(data);
    },
    updater:function(data) {
      return data;
    },
    afterSelect:function(data){
      console.log(data);
    }
  });
```
Is there a fix or a workaround for this?

**Comment by 0s1s** [comment]
Is this library dead?

I just noticed that the above code works on all browsers I mentioned above, IF those browsers are on a desktop. However, if those browsers are on a laptop (at least the one I'm using anyway), then the issue is still present. I don't understand why the above code would work on a desktop and not a laptop. Both desktop and laptop are using the exact same version browser. I even created a new profile in the browsers to ensure that it wasn't a setting getting in the way, but the problem still exists even with the new profiles.

Can someone shed some light on this issue for me? Please?

**Comment by pikot** [comment]
look at  issue - https://github.com/bassjobsen/Bootstrap-3-Typeahead/issues/321 
i think, it is solution

**Comment by pikot** [comment]
ups) no, not solution)

**Comment by 0s1s** [comment]
Thanks pikot, yeah, I already found that thread. I've combed over all of the issues and nothing I've found is close. I don't understand why this library works on a desktop but not on a laptop running the exact same version browsers (Opera and Firefox). I've never seen anything quite like it before.

**Comment by pikot** [comment]
i have same problem or look like yours, i'm not proficient in this code in js, but try to debug =) looks like not run  "mouseup" that inited in mousedown. 
in some sources i saw mention using pointerup if window.PointerEvent is true, but - it is not work too :'(

**Comment by pikot** [comment]
look like floating bug)) 

how i debug, change some code and add warns in funcs, e.g.
```
    mousedown: function (e) {
      this.mouseddown = true;
      var mouseup = function(e){
        // IE won't fire this, but FF and Chrome will so we reset our flag for them here
        console.log("mouseup ");
        this.mouseddown = false;
      };

      if (window.PointerEvent) {
          console.log("pointer event");
          this.$menu.one('pointerup', mouseup.bind(this));
      }
      else {
          console.log("mouse event");
          this.$menu.one('mouseup', mouseup.bind(this));
      }
      console.log("mousedown " + mouseup);
    },
```
some times i saw in console

> pointer event
> bootstrap3-typeahead.js:641 mousedown function (e){
>         // IE won't fire this, but FF and Chrome will so we reset our flag for them here
>         console.log("mouseup ");
>         this.mouseddown = false;
>       }
> bootstrap3-typeahead.js:574 blur, mousedover  true, mouseddown = true, shown = true
> bootstrap3-typeahead.js:620 mouseup 
> bootstrap3-typeahead.js:592 click 

and all work fine,
and some times i saw only 

> pointer event
> bootstrap3-typeahead.js:641 mousedown function (e){
>         // IE won't fire this, but FF and Chrome will so we reset our flag for them here
>         console.log("mouseup ");
>         this.mouseddown = false;
>       }
> bootstrap3-typeahead.js:574 blur, mousedover  true, mouseddown = true, shown = true

**Comment by 0s1s** [comment]
Hi pikot, thanks for the direction. I'm sorry for the late response, but I just got back into town. I'll give this a try and report back what I find. Thank you!

**Comment by 0s1s** [comment]
Pikot, the error I'm getting in the console when the Typeahead input gets focus is: "TypeError: (intermediate value).apply is not a function (jquery.min.js:3:7485)". Also, again, I'm not very good with JavaScript and honestly don't even know where to start debugging this issue. Other than this error, there is absolutely nothing else in the console that indicates there is an issue. I tried adding the code in the mousedown function that you posted above and get nothing in the console. I added a simple alert() in the mousedown function and it never fires. I'm lost. I may just try and find a different library to use. This one seems to be dead.

**Comment by pikot** [comment]
> TypeError: (intermediate value).apply is not a function (jquery.min.js:3:7485)

i think that is root of your problem. You see this topic https://github.com/standard/standard/issues/373 ? maybe in can help you?

**Comment by pikot** [comment]
hm...  i try your example from first message of topic, and all  ok :\
do you run your example on clean page without any other js lib? only jquery and typeahead? 

ps.  perhaps i found reason of my bug, i  miss setting  css  - .typeahead { z-index: 1051; }, and on old chrome now work good, after some hours test on new version chrome

**Comment by 0s1s** [comment]
Thanks for the link pikot, I actually found that solution on Stack Overflow and tried it weeks ago. That didn't solve my issue either. I tested just the bare bones typeahead code it in a clean HTML page outside of my application and the issue still exists on my laptop. This is why I initially submitted the bug report. If the code won't work on the laptop without the possibility of anything else getting in the way, then I have to conclude that there is a bug in the typeahead library. I've honestly never seen anything like this before and I've been developing for over 14 years. Something that works on a desktop but not a laptop, running the exact same version browser, is mind boggling.

**Comment by vince2corte** [comment]
Same problem with Firefox 48 (OS X). Solved by using the non "min" version.
Strange...
Btw : it works with FF, Chrome, IE 11 and Edge

**Comment by 0s1s** [comment]
@vince2corte, thanks Vince. As it turns out, this issue still exists for me, so I can't wait to try this, Thank you!

**Comment by AppLetico** [comment]
Anyone solve this problem?  I am seeing the same issue...

What I see happening is 2 events are firing when you select an item.
Event 1: selects the first value on the list
Event 2: selects the value you actually clicked on

Looks like a bug in the code.  Any maintainers who can look?

**Comment by 0s1s** [comment]
Nope, I think this is still open. And, I believe the library has been dead for some time.

I opted to use a different library, but THIS time, I made sure that it was being maintained. I'm sorry I don't have an answer for you on the bugs, but I **can** refer you to use Bootstrap Combo Box plugin.

See here for more (**does** work with Bootstrap 4): https://www.jqueryscript.net/form/Bootstrap-Combo-Box-Typeahead-Plugin-jQuery.html.

**Comment by AppLetico** [comment]
I was able to workaround the issue by using the afterSelect option.

$("#inputRelated").typeahead({ 
					source: data,
					changeInputOnSelect: false,
					changeInputOnMove: false,
					afterSelect: addRelated
				});

function addRelated(item){
				console.log(item);
			}

**Comment by 0s1s** [comment]
I don't remember exactly what the settings I had were at,  but I never could get the thing to work on any of my laptops. There are actually a couple different issues being talked about in this thread, but my main issues were that I could not get typeahead to detect a mouse click on the dropdown. This behavior only shows up on laptops. It's the weirdest thing. Both laptop and desktops (desktops work fine) running the exact same current versions of Firefox, Opera and Chrome, will not detect onmousedown events. I was able however to use the keyboard arrows up and down to maneuver the selection. Note however, IF YOU FIRST CLICK and then try and use the arrow keys, it's like you're hitting a preventDefault; nothing happens and you have to refresh the screen to try again. Again, this behavior was only present on laptop machines and this library worked great on desktops.

**Comment by vince2corte** [comment]
Well, it seems that the last Firefox has broken my solution, so I found another way to fix it.
First, my control is feeded with an AJAX/JSON query, so I used the code given as an example.
Then I looked at the event sent to my "change" function, and "bingo", the first one is different. I think it is fired by jQuery itself.
So I just found something different betwen the two events, and made a test like that :

my_control.change(function(e) {
        if (typeof e.originalEvent != "undefined") { // my new test
            return true;
        }
	var current = $(this).typeahead("getActive");
	if (current) {
		if (current.name.toLowerCase() == $(this).val().toLowerCase()) {
			$("#'.$name.'").val(current.id);
			do_my_job();
		}
		else {
			$("#'.$name.'").val("");
		}
	}
	else {
		$("#'.$name.'").val("");
	}
});

That's all ! Not very elegant, but it works with all my browsers. And the original source code is not changed.

**Comment by NormandoHall** [comment]
For those are crying with this issue, I fixed adding two lines in `mouseenter` function, for update the `active` value:

```js
mouseenter: function (e) {
  this.mousedover = true;
  this.$menu.find('.active').removeClass('active');
  $(e.currentTarget).addClass('active');
},
```

becomes:

```js
mouseenter: function (e) {
  this.mousedover = true;
  this.$menu.find('.active').removeClass('active');
  $(e.currentTarget).addClass('active');

  // console.log($(e.currentTarget).data('value'))
  this.$element.val(this.displayText($(e.currentTarget).data('value')))
  this.$element.data('active', $(e.currentTarget).data('value'));
},
```
